src/seq_mv/csr_spmv_device.c

```diff
@@ -101,75 +70,100 @@ hypre_csr_v_k_shuffle(hypre_DeviceItem &item,
          grid_row_id = grid_group_id;
       }
 
-      if (grid_group_id < n && group_lane < 2)
+      if (grid_group_id < nrows && group_lane < 2)
       {
          p = read_only_load(&d_ia[grid_row_id + group_lane]);
       }
       q = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 1, K);
       p = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 0, K);
 
-      T sum = 0.0;
-#if VERSION == 1
+      T sum[NV] = {T(0)};
+#if HYPRE_SPMV_VERSION == 1
 #pragma unroll 1
-      for (p += group_lane; warp_any_sync(item, HYPRE_WARP_FULL_MASK, p < q); p += K * 2)
+      for (p += group_lane; p < q; p += K * 2)
       {
-         if (p < q)
+         HYPRE_SPMV_ADD_SUM(p)
+         if (p + K < q)
          {
-            SPMV_ADD_SUM(p)
-            if (p + K < q)
-            {
-               SPMV_ADD_SUM((p + K))
-            }
+            HYPRE_SPMV_ADD_SUM((p + K))
          }
       }
-#elif VERSION == 2
+#elif HYPRE_SPMV_VERSION == 2
 #pragma unroll 1
       for (p += group_lane; warp_any_sync(item, HYPRE_WARP_FULL_MASK, p < q); p += K)
       {
          if (p < q)
          {
-            SPMV_ADD_SUM(p)
+            HYPRE_SPMV_ADD_SUM(p)
          }
       }
 #else
 #pragma unroll 1
       for (p += group_lane;  p < q; p += K)
       {
-         SPMV_ADD_SUM(p)
+         HYPRE_SPMV_ADD_SUM(p)
       }
 #endif
+
       // parallel reduction
 #pragma unroll
       for (HYPRE_Int d = K / 2; d > 0; d >>= 1)
       {
-         sum += __shfl_down_sync(HYPRE_WARP_FULL_MASK, sum, d);
+#pragma unroll
```

Comment:
Have you done any performance testing?  Especially when NV=1 compared to the current code?

Reply:
Hi Rob, I haven't seen any overhead in the new code when NV=1. I do have some performance results, but I need to organize them... I'll do that and post some plots in this PR for documenting. Thanks!

Reply:
Performance results added here: https://github.com/hypre-space/hypre/pull/694#issuecomment-1199923641